src/Std/Internal/Async/Basic.lean

```diff
@@ -13,21 +13,231 @@ namespace Internal
+
+## Relation
+
+Concrete units of work that can be executed within these contexts.
+
+These types are related by two functions in the type classes `MonadAsync` and `MonadAwait`: `async`
+and `await`. The `async` function extracts a concrete asynchronous task from a computation inside
+the monadic context. In effect, it makes the computation run in the background and returns a
+task handle that can be awaited later. On the other side, the `await` function takes a task and
+re-inserts it into the monadic context, allowing the result to be composed using the monadic bind.
+This relation between `async` and `await` enables precise control over when a computation starts and
+when its result is actually used. You can spawn multiple asynchronous tasks using `async`, continue
+with other operations, and later rejoin the computation flow by awaiting their results.
+
+These functions should not be used directly. Instead, prefer higher-level combinators such as
+`race`, `raceAll`, `concurrently`, `background` and `concurrentlyAll`. The best way to think about
+how to write your async code, it to avoid using the concurrent units of work, and only use it when integrating
+with non async code that uses them.
+
+-/
+
+/--
+Typeclass for monads that can "await" a computation of type `t α` in a monad `m` until the result is
+available.
+-/
+class MonadAwait (t : Type → Type) (m : Type → Type) extends Monad m where
+  /--
+  Awaits the result of `t α` and returns it inside the `m` monad.
+  -/
+  await : t α → m α
+
+/--
+Represents monads that can launch computations asynchronously of type `t` in a monad `m`.
+-/
+class MonadAsync (t : Type → Type) (m : Type → Type) extends Monad m where
+  /--
+  Starts an asynchronous computation in another monad.
+  -/
+  async : m α → m (t α)
+
+instance [Monad m] [MonadAwait t m] : MonadAwait t (StateT n m) where
+  await := liftM (m := m) ∘ MonadAwait.await
+
+instance [Monad m] [MonadAwait t m] : MonadAwait t (ExceptT n m) where
+  await := liftM (m := m) ∘ MonadAwait.await
+
+instance [Monad m] [MonadAwait t m] : MonadAwait t (ReaderT n m) where
+  await := liftM (m := m) ∘ MonadAwait.await
+
+instance [MonadAwait t m] : MonadAwait t (StateRefT' s n m) where
+  await := liftM (m := m) ∘ MonadAwait.await
+
+instance [MonadAwait t m] : MonadAwait t (StateT s m) where
+  await := liftM (m := m) ∘ MonadAwait.await
+
+instance [MonadAsync t m] : MonadAsync t (ReaderT n m) where
+  async p := MonadAsync.async ∘ p
+
+instance [MonadAsync t m] : MonadAsync t (StateRefT' s n m) where
+  async p := MonadAsync.async ∘ p
+
+instance [Functor t] [inst : MonadAsync t m] : MonadAsync t (StateT s m) where
+  async p := fun s => do
+    let t ← inst.async (p s)
+    pure (t <&> Prod.fst, s)
+
+
+/--
+A `Task` that may resolve to either a value of type `α` or an error value of type `ε`.
+-/
+def ETask (ε : Type) (α : Type u) : Type u := Task (Except ε α)
```

Comment:
Perhaps this could be an `abbrev` rather than a `def`? It seems that `ETask` and `Task (Except ε α)` are pretty interchangeable (`ETask` just being shorter and having nicer dot notation). 

Also, did you consider implementing `ETask` as `ExceptT ε (Task α)`?

Reply:
With `ExceptT`, you could potentially get many of the monadic operations, instances, and lemmas for free.